spec/models/penalty_decay_period_submission_rule_spec.rb

```diff
@@ -65,7 +33,56 @@
     teardown do
       destroy_repos
     end
-
+    context 'when the student did not submit any files' do
+      # Before the due_date and before collection_time
+      describe 'when the assignment is collected before the due date' do
+        before :each do
+          @submission = create(:submission, grouping: @grouping, is_empty: true)
+        end
+        it 'should not apply submission rule' do
+          pretend_now_is(Time.now + 1.days) do
+            expect(Time.now).to be < @assignment.due_date
+            expect(Time.now).to be < @assignment.submission_rule.calculate_collection_time
+            result = @submission.get_latest_result
+            expect(result).not_to be_nil
+            expect(result.extra_marks).to be_empty
+            expect(result.get_total_extra_percentage).to eq 0
+          end
+        end
+      end
+      # After the due_date and before collection_time
+      describe 'when the assignment is collected after the due date' do
+        before :each do
+          @group = create(:group)
+          @grouping = create(:grouping, group: @group)
+          @assignment = @grouping.assignment
+          @rule = PenaltyDecayPeriodSubmissionRule.new
+          @assignment.replace_submission_rule(@rule)
+          PenaltyDecayPeriodSubmissionRule.destroy_all
+          @rule.save
+          @submission = create(:submission, grouping: @grouping, is_empty: true)
+          # The instructor sets up the course...
+          @assignment.due_date = Time.now + 2.days
+          # Add two 24 hour penalty decay periods
+          # Overtime begins in two days.
+          add_period_helper(@assignment.submission_rule, 24, 10, 12)
+          add_period_helper(@assignment.submission_rule, 24, 10, 12)
+          # Collection date is in 4 days.
+          @assignment.save
+          @grouping.create_grouping_repository_folder
+        end
+        it 'should not apply submission rule' do
```

Comment:
@ANUSHIYAPRIYA I really like your use of shared examples for the grace tokens. It seems that you can also use a new shared example for the other two submission rules (for penalty periods) since both rely on checking that there are no extra marks created.